### PR TITLE
MOB-4628 fix payment unit tests

### PR DESCRIPTION
**Pull request by dianakarh**
## JIRA
[MOB-4628](https://karhoo.atlassian.net/browse/MOB-4628)

## DESCRIPTION
[Create a description of what you have changed in this branch]

## Types of changes

What types of changes does your code introduce to the project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix 
- [ ] New feature 
- [ ] Tech Debt
- [ ] Breaking change 
- [ ] Documentation Update
- [ ] Other: *define what type of change it is*

## CHECKLIST

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. This is simply a reminder of what we are going to look for before merging code._
Please review the guidelines for contributing to this repository.

- [ ] All work relating to the ticket is complete (Code complete/Acceptance Criteria met)
- [ ] Lint and unit tests pass locally (if appropriate)
- [ ] Added tests that prove the fix is effective or that feature works (if appropriate)
- [ ] Documentation (if appropriate)

## SCREENSHOTS
[Add screenshots if any for changes you have made to the App/UI]

[MOB-4628]: https://karhoo.atlassian.net/browse/MOB-4628?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: KarhooUISDKTestUtils/Mocks/Common/MockCheckoutPaymentWorker.swift [hunk 1]
```diff
@@ -36,6 +36,10 @@ public class MockCheckoutPaymentWorker: CheckoutPaymentWorker {
         getPaymentNonceCalled = true
         completion(getPaymentNonceResult)
     }
+    
+    public func resetStoredPaymentNonce() {
```
Comment (AleksanderWedrychowskiKarhoo) [inline KarhooUISDKTestUtils/Mocks/Common/MockCheckoutPaymentWorker.swift:40]: Wouldn't 'clear' a more precise name for this behavior?

File: KarhooUISDK/Screens/CheckoutScreen/Checkout/Workers/KarhooCheckoutPaymentWorker.swift [hunk 2]
```diff
@@ -68,13 +69,16 @@ final class KarhooCheckoutPaymentWorker: CheckoutPaymentWorker {
     func getStoredPaymentNonce() -> Nonce? {
         paymentNonce
     }
+    
+    func resetStoredPaymentNonce() {
```
Comment (AleksanderWedrychowskiKarhoo) [inline KarhooUISDK/Screens/CheckoutScreen/Checkout/Workers/KarhooCheckoutPaymentWorker.swift:73]: Wouldn't 'clear' a more precise name for this behavior?

File: KarhooUISDK/Screens/CheckoutScreen/Checkout/Workers/KarhooCheckoutBookingWorker.swift [hunk 3]
```diff
@@ -327,33 +287,22 @@ final class KarhooCheckoutBookingWorker: CheckoutBookingWorker {
...
         } else if let error = result.getErrorValue() {
             reportBookingFailure(message: error.message, correlationId: result.getCorrelationId())
+            
+            // Show the error to the user
             stateSubject.send(.failure(result.getErrorValue() ?? ErrorModel.unknown()))
+            
+            // Adyen pre-auth usually fails before booking,
+            // while for Braintree we still get a nonce and the payment is refused after the booking call is made.
+            // For this scenario we need to reset the nonce and let the user restart the payment flow from scratch
+            if error.type == .couldNotBookTripPaymentPreAuthFailed {
```
Comment (AleksanderWedrychowskiKarhoo) [inline KarhooUISDK/Screens/CheckoutScreen/Checkout/Workers/KarhooCheckoutBookingWorker.swift:303]: What would happen if we clear a nonce if any error occurs?

**Review by AleksanderWedrychowskiKarhoo — CHANGES_REQUESTED** [review]

File: KarhooUISDK/Screens/CheckoutScreen/Checkout/Workers/KarhooCheckoutBookingWorker.swift (hunk 3 above)
Comment (dianakarh) [reply KarhooUISDK/Screens/CheckoutScreen/Checkout/Workers/KarhooCheckoutBookingWorker.swift:303]: The rest of the errors in the enum are not payment related judging by their names, so if we were to clear the nonce in their case we'd require the user to re-authorise the amount and we'd block the same amount as many times as there are attempts to book. Even if we unblock the amounts later, it's still bad UX in my opinion and could lead to confusion and mistrust.

**Review by AleksanderWedrychowskiKarhoo — APPROVED** [review]

**Review by BartlomiejSopalaKarhoo — APPROVED** [review]

**Pull request merged**